File: EngineInterface.Test/Engine/Core/TimerTests.cs
using System.Threading;
using EngineInterface.Managed.Engine.Core;
using Xunit;

namespace EngineInterface.Tests.Engine.Core
{
    public class TimerTests
    {
        [Fact]
        public void Constructor_ShouldStartTimerInUsableState()
        {
            var timer = new TimerWrapper();

            var elapsed = timer.TimeElapsed();

            Assert.InRange(elapsed, 0.0, 0.1);
        }

        [Fact]
        public void Reset_ShouldRestartElapsedTime()
        {
            var timer = new TimerWrapper();

            Thread.Sleep(20);
            var beforeReset = timer.TimeElapsed();

            timer.Reset();
            var afterReset = timer.TimeElapsed();

            Assert.True(beforeReset > 0.0);
            Assert.InRange(afterReset, 0.0, 0.05);
        }

        [Fact]
        public void TimeElapsed_ShouldIncreaseOverTime()
        {
            var timer = new TimerWrapper();

            var first = timer.TimeElapsed();
            Thread.Sleep(20);
            var second = timer.TimeElapsed();

            Assert.True(second > first);
        }

        [Fact]
        public void Tick_ShouldReturnPositiveDeltaAfterDelay()
        {
            var timer = new TimerWrapper();

            Thread.Sleep(20);
            var delta = timer.Tick();

            Assert.True(delta > 0.0);
            Assert.InRange(delta, 0.0, 0.2);
        }

        [Fact]
        public void ConsecutiveTick_ShouldMeasureSeparateIntervals()
        {
            var timer = new TimerWrapper();

            Thread.Sleep(15);
            var firstDelta = timer.Tick();

            Thread.Sleep(30);
            var secondDelta = timer.Tick();

            Assert.True(firstDelta > 0.0);
            Assert.True(secondDelta > 0.0);
            Assert.True(secondDelta > firstDelta);
        }

        [Fact]
        public void Tick_ShouldUpdatePreviousReferenceWithoutResettingElapsedTime()
        {
            var timer = new TimerWrapper();

            Thread.Sleep(20);
            timer.Tick();

            Thread.Sleep(15);
            var secondTick = timer.Tick();
            var elapsed = timer.TimeElapsed();

            Assert.True(secondTick > 0.0);
            Assert.True(elapsed >= secondTick);
        }

        [Fact]
        public void Reset_AfterTick_ShouldRestartElapsedAndTickBaselines()
        {
            var timer = new TimerWrapper();

            Thread.Sleep(20);
            timer.Tick();

            timer.Reset();

            Thread.Sleep(15);
            var deltaAfterReset = timer.Tick();
            var elapsedAfterReset = timer.TimeElapsed();

            Assert.True(deltaAfterReset > 0.0);
            Assert.True(elapsedAfterReset > 0.0);
            Assert.InRange(deltaAfterReset, 0.0, 0.2);
            Assert.InRange(elapsedAfterReset, 0.0, 0.2);
        }
    }
}

File: EngineInterface.NativeTests/Code/Engine/Core/TimerTests.cpp
#include "CppUnitTest.h"

#include <Engine/Core/Timer.h>
#include <thread>
#include <chrono>

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace Engine
{
    namespace Core
    {
        TEST_CLASS(TimerTests)
        {
        public:

            // ======================================================
            // Timer Updates
            // ======================================================

            TEST_METHOD(Reset_ShouldStartTimerInUsableState)
            {
                Timer timer;
                timer.Reset();

                double elapsed = timer.TimeElapsed();

                Assert::IsTrue(elapsed >= 0.0 && elapsed <= 0.1);
            }

            TEST_METHOD(Reset_ShouldRestartElapsedTime)
            {
                Timer timer;
                timer.Reset();

                std::this_thread::sleep_for(std::chrono::milliseconds(20));
                double beforeReset = timer.TimeElapsed();

                timer.Reset();
                double afterReset = timer.TimeElapsed();

                Assert::IsTrue(beforeReset > 0.0);
                Assert::IsTrue(afterReset >= 0.0 && afterReset <= 0.05);
            }

            // ======================================================
            // Behaviour
            // ======================================================

            TEST_METHOD(TimeElapsed_ShouldIncreaseOverTime)
            {
                Timer timer;
                timer.Reset();

                double first = timer.TimeElapsed();
                std::this_thread::sleep_for(std::chrono::milliseconds(20));
                double second = timer.TimeElapsed();

                Assert::IsTrue(second > first);
            }

            // ======================================================
            // Timer Updates
            // ======================================================

            TEST_METHOD(Tick_ShouldReturnPositiveDeltaAfterDelay)
            {
                Timer timer;
                timer.Reset();

                std::this_thread::sleep_for(std::chrono::milliseconds(20));
                double delta = timer.Tick();

                Assert::IsTrue(delta > 0.0);
                Assert::IsTrue(delta >= 0.0 && delta <= 0.2);
            }

            TEST_METHOD(ConsecutiveTick_ShouldMeasureSeparateIntervals)
            {
                Timer timer;
                timer.Reset();

                std::this_thread::sleep_for(std::chrono::milliseconds(15));
                double firstDelta = timer.Tick();

                std::this_thread::sleep_for(std::chrono::milliseconds(30));
                double secondDelta = timer.Tick();

                Assert::IsTrue(firstDelta > 0.0);
                Assert::IsTrue(secondDelta > 0.0);
                Assert::IsTrue(secondDelta > firstDelta);
            }

            TEST_METHOD(Tick_ShouldUpdatePreviousReferenceWithoutResettingElapsedTime)
            {
                Timer timer;
                timer.Reset();

                std::this_thread::sleep_for(std::chrono::milliseconds(20));
                timer.Tick();

                std::this_thread::sleep_for(std::chrono::milliseconds(15));
                double secondTick = timer.Tick();
                double elapsed = timer.TimeElapsed();

                Assert::IsTrue(secondTick > 0.0);
                Assert::IsTrue(elapsed >= secondTick);
            }

            TEST_METHOD(Reset_AfterTick_ShouldRestartElapsedAndTickBaselines)
            {
                Timer timer;
                timer.Reset();

                std::this_thread::sleep_for(std::chrono::milliseconds(20));
                timer.Tick();

                timer.Reset();

                std::this_thread::sleep_for(std::chrono::milliseconds(15));
                double deltaAfterReset = timer.Tick();
                double elapsedAfterReset = timer.TimeElapsed();

                Assert::IsTrue(deltaAfterReset > 0.0);
                Assert::IsTrue(elapsedAfterReset > 0.0);
                Assert::IsTrue(deltaAfterReset <= 0.2);
                Assert::IsTrue(elapsedAfterReset <= 0.2);
            }
        };
    }
}